In [1]:
# Set up packages for lecture. Don't worry about understanding this code, but
# make sure to run it if you're following along.
import numpy as np
import babypandas as bpd
import pandas as pd
from matplotlib_inline.backend_inline import set_matplotlib_formats
%reload_ext pandas_tutor
%set_pandas_tutor_options {'projectorMode': True}
set_matplotlib_formats("svg")

np.set_printoptions(threshold=20, precision=2, suppress=True)
pd.set_option("display.max_rows", 7)
pd.set_option("display.max_columns", 8)
pd.set_option("precision", 2)

# Lecture 6 Challenge Question Solutions

## DSC 10, Summer 2022

## Challenge Problems

We won't cover these problems in class, but they're here for you to practice with some harder examples. To access the solutions, you'll need to watch [this solution walkthrough video](https://youtu.be/xg7rnjWnZ48).

Before accessing the solutions, **make sure to try these problems on your own** – they're great prep for homeworks and projects!

In [2]:
imdb = bpd.read_csv('data/imdb.csv').set_index('Title').sort_values(by='Rating')
imdb

Votes  Rating  Year  Decade
Title                                                  
Akira                       91652     8.0  1988    1980
Per un pugno di dollari    124671     8.0  1964    1960
Guardians of the Galaxy    527349     8.0  2014    2010
...                           ...     ...   ...     ...
The Godfather: Part II     692753     9.0  1974    1970
The Shawshank Redemption  1498733     9.2  1994    1990
The Godfather             1027398     9.2  1972    1970

[250 rows x 4 columns]

### Question: How many years have more than 3 movies rated above 8.5?

In [3]:
# BEGIN SOLUTION
good_movies_per_year = imdb[imdb.get('Rating') > 8.5].groupby('Year').count()
good_movies_per_year[good_movies_per_year.get('Votes') > 3].shape[0]
# END SOLUTION

1

#### Aside: Using `.sum()` on a boolean array

- Summing a boolean array gives a count of the number of `True` elements because Python treats `True` as 1 and `False` as 0. 
- Can you use that fact here?

In [4]:
(good_movies_per_year.get('Votes') > 3).sum() 

1

### Question: Out of the years with more than 3 movies, which had the highest average rating?

In [5]:
# BEGIN SOLUTION
more_than_3_ix = imdb.groupby('Year').count().get('Votes') > 3
imdb.groupby('Year').mean()[more_than_3_ix].sort_values(by='Rating').index[-1]
# END SOLUTION

1994

### Question: Which year had the longest movie titles, on average?

**Hint:** Use `.str.len()` on the column or index that contains the names of the movies.

In [6]:
# BEGIN SOLUTION
(
    imdb.assign(title_length=imdb.index.str.len())
    .groupby('Year').mean()
    .sort_values(by='title_length')
    .index[-1]
)
# END SOLUTION
imdb[imdb.get('Year') == 1964] # SOLUTION NO PROMPT

Votes  Rating  Year  \
Title                                                                      
Per un pugno di dollari                             124671     8.0  1964   
Dr. Strangelove or: How I Learned to Stop Worry...  309141     8.5  1964   

                                                    Decade  
Title                                                       
Per un pugno di dollari                               1960  
Dr. Strangelove or: How I Learned to Stop Worry...    1960

### Question: What is the average rating of movies from years that had at least 3 movies in the Top 250?

In [7]:
# BEGIN SOLUTION
# A Series of Trues and Falses; True when there were at least 3 movies on the list from that year
more_than_3_ix = imdb.groupby('Year').count().get('Votes') > 3

# The sum of the ratings of movies from years that had at least 3 movies on the list
total_rating = imdb.groupby('Year').sum()[more_than_3_ix].get('Rating').sum()

# The total number of movies from years that had at least 3 movies on the list
count = imdb.groupby('Year').count()[more_than_3_ix].get('Rating').sum()

# The correct answer
average_rating = total_rating / count

# Close, but incorrect – doesn't account for the fact that different years have different numbers of movies on the list
close_but_wrong = imdb.groupby('Year').mean()[more_than_3_ix].get('Rating').mean()
# END SOLUTION

In [8]:
average_rating

8.262576687116566